File: LeetCode/0368. Largest Divisible Subset/solution.py
"""
48 / 48 test cases passed.
Runtime: 388 ms
Memory Usage: 15.1 MB
"""
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        nums = sorted(nums)
        largest_subset = []
        dp = [[num] for num in nums]
        for i in range(len(nums)):
            for j in range(i):
                if nums[i] % nums[j] == 0 and len(dp[j]) + 1 > len(dp[i]):
                    dp[i] = dp[j] + nums[i:i+1]
            if len(largest_subset) < len(dp[i]):
                largest_subset = dp[i]
        return largest_subset

File: LeetCode/0368. Largest Divisible Subset/solution.cpp
/**
 * 48 / 48 test cases passed.
 * Runtime: 40 ms
 * Memory Usage: 8.5 MB 
 */
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<int> dp(nums.size(), 1);
        int maxVal = 0, maxSize = 1;
        for (int i = 1; i < nums.size(); ++ i) {
            for (int j = 0; j < i; ++ j) {
                if (nums[i] % nums[j] == 0) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
            if (maxSize < dp[i]) {
                maxSize = dp[i];
                maxVal = nums[i];
            }
        }
        vector<int> ans;
        for (int i = nums.size() - 1; i >= 0; -- i) {
            if (dp[i] == maxSize && 
            maxSize && maxVal % nums[i] == 0) {
                ans.push_back(nums[i]);
                maxVal = nums[i];
                maxSize--;
            }
        }
        return ans;
    }
};
